### PR TITLE
Loosen error message matching for IMDS 403 responsesx

### PR DESCRIPTION
**Pull request by anannya03**
# Description

[Task](https://github.com/Azure/azure-sdk-for-java/issues/38358)

Adjusted error handling logic to accommodate variations in the IMDS 403 error message observed in Docker. Previously, the logic matched exact strings. It now supports both:

"A socket operation was attempted to an unreachable network"

"A socket operation was attempted to an unreachable host"

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/identity/azure-identity/src/test/java/com/azure/identity/implementation/ImdsRetryStrategyTest.java [hunk 1]
```diff
@@ -50,13 +50,29 @@ public void testShouldRetry() {
...
+            "Imds Retry Strategy should retry on 403 status with unreachable network message.");
+
+        MockHttpResponse responseStatus403Host = new MockHttpResponse(null, 403);
+        responseStatus403Host.addHeader("ResponseMessage", "A socket operation was attempted to an unreachable host");
+        Assertions.assertTrue(imdsRetryStrategy.shouldRetry(responseStatus403Host),
+            "Imds Retry Strategy should retry on 403 status with unreachable host message.");
+
+        MockHttpResponse responseStatus403NotSuccesful = new MockHttpResponse(null, 403);
+        responseStatus403NotSuccesful.addHeader("ResponseMessage", "Access denied");
+        Assertions.assertFalse(imdsRetryStrategy.shouldRetry(responseStatus403NotSuccesful),
```
Comment (alzimmermsft) [inline sdk/identity/azure-identity/src/test/java/com/azure/identity/implementation/ImdsRetryStrategyTest.java]: Given this is the only test to assertFalse, would be nice to make it its own test.

File: sdk/identity/azure-identity/src/test/java/com/azure/identity/implementation/ImdsRetryStrategyTest.java [hunk 2]
```diff
@@ -50,13 +50,29 @@ public void testShouldRetry() {
...
+        MockHttpResponse responseStatus403Network = new MockHttpResponse(null, 403);
+        responseStatus403Network.addHeader("ResponseMessage",
+            "A socket operation was attempted to an unreachable network");
+        Assertions.assertTrue(imdsRetryStrategy.shouldRetry(responseStatus403Network),
+            "Imds Retry Strategy should retry on 403 status with unreachable network message.");
+
+        MockHttpResponse responseStatus403Host = new MockHttpResponse(null, 403);
+        responseStatus403Host.addHeader("ResponseMessage", "A socket operation was attempted to an unreachable host");
+        Assertions.assertTrue(imdsRetryStrategy.shouldRetry(responseStatus403Host),
+            "Imds Retry Strategy should retry on 403 status with unreachable host message.");
```
Comment (alzimmermsft) [inline sdk/identity/azure-identity/src/test/java/com/azure/identity/implementation/ImdsRetryStrategyTest.java]: Would be nice to make this entire test into a `ParameterizedTest` given we're testing multiple different cases, and we could introduce a change that breaks cases 1, 2, and 3, but debugging through that would require multiple runs as it would fail fast.

`ParameterizedTest`s are good for shared logic but different cases, such as this, where each case is its own run.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/ImdsRetryStrategy.java [hunk 3]
```diff
@@ -58,8 +58,12 @@ public boolean shouldRetry(HttpResponse httpResponse) {
                 return false;
             }
             if (statusCode == 403) {
-                return httpResponse.getHeaderValue("ResponseMessage")
-                    .contains("A socket operation was attempted to an unreachable network");
+                String headerValue = httpResponse.getHeaderValue("ResponseMessage");
+                if (headerValue != null) {
+                    return headerValue.contains("A socket operation was attempted to an unreachable network")
+                        || headerValue.contains("A socket operation was attempted to an unreachable host");
```
Comment (chlowell) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/ImdsRetryStrategy.java]:
```suggestion
                    return headerValue.contains("A socket operation was attempted to an unreachable");
```

File: sdk/identity/azure-identity/src/test/java/com/azure/identity/implementation/ImdsRetryStrategyTest.java (hunk 1 above)
Comment (anannya03) [reply sdk/identity/azure-identity/src/test/java/com/azure/identity/implementation/ImdsRetryStrategyTest.java]: I’ve added another test case where the ResponseMessage header value is null, and also noticed that the 404 status case results in assertFalse as well. That brings the total to three assertFalse scenarios.

For now, I have updated the parameterizedTest to include a expectedRetry boolean to handle both true and false cases within the same test. Let me know if you'd prefer these to be split into separate tests for clarity.

**Comment by g2vinay** [comment]
Run 'mvn spotless:apply'  locally and then check in those changes to fix build issues.
